ui/dashboards/src/templating/core.tsx

```diff
@@ -0,0 +1,269 @@
+
+type TemplateVariableSrv = {
+  variableDefinitions: VariableDefinitions;
+  variableState: VariablesState;
+  setVariableValue: (variableName: VariableName, value: VariableValue) => void;
+  loadTemplateVariable: (name: VariableName) => Promise<void>;
+};
+
+const TemplateVariableSrvContext = createContext<ReturnType<typeof createTemplateVariableSrvStore> | undefined>(
+  undefined
+);
+function useTemplateVariableSrvContext() {
+  const context = useContext(TemplateVariableSrvContext);
+  if (!context) {
+    throw new Error('TemplateVariableSrvContext not initialized');
+  }
+  return context;
+}
+
+export function useTemplateVariableValues(variableNames?: string[]) {
+  const store = useTemplateVariableSrvContext();
+  const state = useStore(
+    store,
+    (s) => {
+      const names = variableNames ?? Object.keys(s.variableState);
+      const vars: VariablesState = {};
+      names.forEach((name) => {
+        const varState = s.variableState[name];
+        if (!varState) {
+          return;
+        }
+        vars[name] = varState;
+      });
+      return vars;
+    },
+    (left, right) => {
+      return JSON.stringify(left) === JSON.stringify(right);
+    }
+  );
+  return state;
+}
+
+export function useTemplateVariable(name: string) {
+  const store = useTemplateVariableSrvContext();
+  return useStore(store, (s) => {
+    const variableState = s.variableState[name];
+    const definition = s.variableDefinitions.find((v) => v.name === name);
+    return {
+      state: variableState,
+      definition,
+    };
+  });
+}
+
+export function useTemplateVariableActions() {
+  const store = useTemplateVariableSrvContext();
+  return useStore(store, (s) => {
+    return {
+      setVariableValue: s.setVariableValue,
+      loadTemplateVariable: s.loadTemplateVariable,
+    };
+  });
+}
+
+export function useTemplateVariableDefinitions() {
+  const store = useTemplateVariableSrvContext();
+  return useStore(store, (s) => s.variableDefinitions);
+}
+
+export function useTemplateVariableSrv() {
+  const store = useTemplateVariableSrvContext();
+  return useStore(store);
+}
+
+function PluginProvider({ children }: { children: React.ReactNode }) {
+  const values = useTemplateVariableValues();
+  return <TemplateVariableContext.Provider value={{ state: values }}>{children}</TemplateVariableContext.Provider>;
+}
+
+interface TemplateVariableSrvArgs {
+  initialVariableDefinitions?: VariableDefinitions;
+}
+
+function createTemplateVariableSrvStore({ initialVariableDefinitions = [] }: TemplateVariableSrvArgs) {
+  const store = createStore<TemplateVariableSrv>()(
+    devtools(
+      immer((set, get) => ({
+        variableState: hydrateTemplateVariableStates(initialVariableDefinitions),
+        variableDefinitions: initialVariableDefinitions,
+
+        // Actions
+        loadTemplateVariable: async (name: VariableName) => {
+          const def = get().variableDefinitions.find((v) => v.name === name) as ListVariableDefinition;
```

Comment:
Why the cast?

Reply:
loadingTemplateVariables is only required for ListVariable since TextVariables don't have options. I actually may need to change this logic when I implement dynamic plugin resolvers.